### PR TITLE
Feat date-functions article

### PR DESCRIPTION
**Pull request by AntonZogk**
Merge request template: please remove the appropriate parts of this template. 

Pre-merge request checklist (to be completed by the one making the request):
 - [ ] I have performed a full review of this code myself.
    - For Python code in PySpark specific sections, all code should have been run in Jupyter notebooks. 
    - For code in sections of the book containing both Python and R code, the page of the book should be constructed as described in the [contributing guide](https://github.com/best-practice-and-impact/ons-spark/blob/main/CONTRIBUTING.md) and converted to a markdown file. 
 - [ ] I have formatted the outputs of code blocks correctly (to match other outputs in the book and in line with the style guide [coming soon])
 - [ ] I have built the book as outlined in the [contributing guide](https://github.com/best-practice-and-impact/ons-spark/blob/main/CONTRIBUTING.md) and confirmed that any additional/modified content is displaying as expected.

Details of this request (such as):
Added guidance on dates in pyspark and sparklyr

Requirements for review (such as):
- Can you also please review that the new chapter is fine being under spark functions ?

### CODE REVIEW

**Review by NathanKelly-ONS — CHANGES_REQUESTED** [review]
Hey Anton, could you include the other files that go along with this please, i.e. the notebook, the r_input and the output file that is created when you run the notebook converter? Just so we can recreate the markdown if needed, and so I can run the code to make sure it all runs correctly :)

**Comment by AntonZogk** [comment]
@NathanKelly-ONS i didn't use the converter for this, i created the markdown from scratch

**Review by NathanKelly-ONS — CHANGES_REQUESTED** [review]
Mostly looks good - I've made some very small changes to the grammar (i.e. there were a few instances with a full stop but no space after it) but aside from that I think the content looks good.

The only suggestion I have is there's an inner join at some point in the code, where you use an inner join to find any animal rescue incidents that occurred on festive days (if you do ctrl + F and search for "inner join" you should find it). 

Although that gets the job done, I don't feel like it's best practice (we want to avoid using joins if at all possible - it's obviously fine with such a small dataset but may be problematic with huge datasets) and it avoids a perfectly good opportunity to talk about filtering by dates (which I think is a very important thing to mention). Could you re-write that section to mention filtering? 

I think some additional parts of the code use the "festive_day" column that is created by using the inner join, i.e. the next section does a "groupby" on the "festive_day" column (so maybe you could add in an additional block of code saying something like "Note: a join here may not be the most efficient method if you want to find all occurrences that happened over festive periods, an alternative way is to use filtering, here's how you would do this: " and then explain how you'd filter for specific dates.

Hope that makes sense!

File: ons-spark/raw-notebooks/date-functions/date-functions.ipynb [hunk 1]
```diff
@@ -395,195 +397,114 @@
    "metadata": {},
    "source": [
     "If only one component of the date is required then there are specific functions: `F.year()`, `F.month()` and `F.dayofmonth()`; in theory these should be more efficient than `F.date_format()`, although less elegant to write if being combined. For `sparklyr` we can achieve the same result by applying `year()`, `month()`, `dayofmonth()` with `mutate()`.\n",
-    " Now create a new DataFrame containing some festive days:"
+    "\n",
+    "In this example we are aiming to analyse the incidents during festive days. There are many ways to approach this problem, ideally we want to leverage the spark narrow transformations if possible, since they are much more efficient compared to wide transformations. \n",
```
Comment (NathanKelly-ONS) [inline ons-spark/raw-notebooks/date-functions/date-functions.ipynb:401]: Can you change this so Spark is capitalised please?

File: ons-spark/raw-notebooks/date-functions/date-functions.ipynb [hunk 2]
```diff
@@ -642,89 +569,89 @@
...
      },
-     "execution_count": 8,
+     "execution_count": 20,
      "metadata": {},
      "output_type": "execute_result"
     }
    ],
    "source": [
-    "festive_rescues.groupBy(\"date_of_call\", \"festive_day\").count().orderBy(\"date_of_call\").limit(10).toPandas()"
+    "festive_rescues.groupBy(\"date_of_call\", \"call_month_day\").count().orderBy(\"call_month_day\",\"date_of_call\").replace(festive_dates).withColumnRenamed(\"call_month_day\",\"festive_day\").limit(10).toPandas()"
```
Comment (NathanKelly-ONS) [inline ons-spark/raw-notebooks/date-functions/date-functions.ipynb:654]: This is a lot to have on a single line, can we enclose it in brackets and go to separate lines? Something like this:

(festive_rescues.groupBy("date_of_call", "call_month_day")
 .count()
 .orderBy("call_month_day","date_of_call")
 .replace(festive_dates)
 .withColumnRenamed("call_month_day","festive_day")
 .limit(10).toPandas())

**Review by NathanKelly-ONS — CHANGES_REQUESTED** [review]
Hey Anton, the changes look good, I just have a couple of minor formatting changes to suggest.

The only other thing is please could you re-run the notebook converter to update the markdown file and the R_input.R script?

**Review by NathanKelly-ONS — CHANGES_REQUESTED** [review]
Looks good - could you please just run the notebook converter and commit the latest markdown file/R_input files? Unfortunately I can't do it myself because Git in DevTest isn't working for me at the moment.

**Review by NathanKelly-ONS — APPROVED** [review]

**Pull request merged**